File: platform/darwin/ios/animax_framework/loader/AnimaXLoaderResponse.m
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#import <AnimaX/AnimaXLoaderProtocol.h>
#import <AnimaX/AnimaXLoaderResponse.h>
#import "AnimaXLoaderResponse+Internal.h"

@interface AnimaXLoaderResponse ()
@property(nonatomic, readwrite) id payload;
@property(nonatomic, readwrite) NSError* error;
@end

NSError* AnimaXErrorMakeWithNSString(NSString* message) {
  return [NSError errorWithDomain:NSCocoaErrorDomain
                             code:-1
                         userInfo:@{NSLocalizedDescriptionKey : message}];
}

AnimaXLoaderPayloadType AnimaXLoaderResponsePayloadTypeForPayload(id payload) {
  if ([payload isKindOfClass:[NSString class]]) {
    return AnimaXLoaderPayloadTypeFilePathNSString;
  } else if ([payload isKindOfClass:[NSData class]]) {
    return AnimaXLoaderPayloadTypeNSData;
  } else {
    return AnimaXLoaderPayloadTypeInvalid;
  }
}

@implementation AnimaXLoaderResponse

- (instancetype)initWithPayload:(id)payload {
  self = [super init];
  if (self) {
    _payload = payload;
    _payloadType = AnimaXLoaderResponsePayloadTypeForPayload(_payload);
    if (_payloadType == AnimaXLoaderPayloadTypeInvalid) {
      _payload = nil;
      _error = AnimaXErrorMakeWithNSString(
          @"Invalid payload: payload must be either NSString* or NSData*.");
    }
  }
  return self;
}

- (instancetype)initWithError:(NSError*)error {
  self = [super init];
  if (self) {
    _payloadType = AnimaXLoaderPayloadTypeInvalid;
    _payload = nil;
    if (error != nil) {
      _error = error;
    } else {
      _error = AnimaXErrorMakeWithNSString(@"Invalid error: error must not be nil.");
    }
  }
  return self;
}

+ (AnimaXLoaderResponse*)responseWithPayload:(nonnull id)payload {
  return [[AnimaXLoaderResponse alloc] initWithPayload:payload];
}

+ (AnimaXLoaderResponse*)responseWithError:(nonnull NSError*)error {
  return [[AnimaXLoaderResponse alloc] initWithError:error];
}

+ (AnimaXLoaderResponse*)responseWithErrorMessage:(NSString*)errorMessage {
  return [[AnimaXLoaderResponse alloc]
      initWithError:[NSError errorWithDomain:@"com.lynx.animax"
                                        code:1
                                    userInfo:@{NSLocalizedDescriptionKey : errorMessage}]];
}

@end

File: animax/platform/android/animax_android/src/main/java/com/lynx/animax/loader/AnimaXLoaderResponse.java
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.animax.loader;

import android.graphics.Bitmap;
import androidx.annotation.NonNull;

public final class AnimaXLoaderResponse<T> {
  public enum Type { STRING_FILE_PATH, BYTE_ARRAY, BITMAP, ERROR }
  private final Type mType;
  private final T mData;

  public static AnimaXLoaderResponse<String> createStringFilePathResponse(
      @NonNull String filePath) {
    return new AnimaXLoaderResponse<>(Type.STRING_FILE_PATH, filePath);
  }

  public static AnimaXLoaderResponse<byte[]> createByteArrayResponse(@NonNull byte[] bytes) {
    return new AnimaXLoaderResponse<>(Type.BYTE_ARRAY, bytes);
  }

  public static AnimaXLoaderResponse<IAnimaXCloseableBitmapReference> createBitmapResponse(
      @NonNull IAnimaXCloseableBitmapReference bitmap) {
    return new AnimaXLoaderResponse<>(Type.BITMAP, bitmap);
  }

  public static AnimaXLoaderResponse<Throwable> createErrorResponse(@NonNull Throwable error) {
    return new AnimaXLoaderResponse<>(Type.ERROR, error);
  }

  public T getData() {
    return mData;
  }

  Type getType() {
    return mType;
  }

  private AnimaXLoaderResponse(Type type, T data) {
    this.mType = type;
    this.mData = data;
  }
}
